### DISCUSSION TITLE
Constructing and monitoring multi-level dynamic dagster jobs

### DISCUSSION THREAD

**Discussion by Phazure** [dagster-io/dagster#7422] [Q&A]
# Use case details
- I would like to set up a dagster job in which the exact DAG would be dynamically generated at runtime based on the contents of an input config. More specifically, the DAG would be comprised of multiple levels of dynamic graphs, where each op could be succeeded by another dynamic graph as shown below.
- On success or fail of each complete DAG job, a slack message report is to be generated and sent, describing the specific ops which succeeded / failed.

![dynamic_job](https://user-images.githubusercontent.com/17210429/163074703-d6eada98-2f0a-4830-8761-9e0555fc36a5.png)

With this use case in mind, was wondering if we could get some additional thoughts on how we could make this work using the following efforts as a starting point, and whether there are any feature improvements that could be implemented to better enable this use case.

# Example config
A slightly contrived example config.
```
ops:
  run_complete_workflow:
    config:
      downstream:
        - run_name: mfd1
          param1: 1
          param2: 2
          downstream:
            - run_name: nav1
              param1: 3
              param2: 4
            - run_name: nav2
              param1: 5
              param2: 6
        - run_name: mfd2
          param1: 10
          param2: 20
          downstream:
            - run_name: nav3
              param1: 30
              param2: 40
            - run_name: nav4
              param1: 50
              param2: 60
        - run_name: mfd3
          param1: 11
          param2: 12
          downstream:
            - run_name: nav5
              param1: 13
              param2: 14
            - run_name: nav6
              param1: 15
              param2: 16
```
# Approaches
## Approach 1
Our first instinct was to leverage nested subgraphs, where we continually pass in the next graph's config. However, we quickly realized that **multiple downstream dynamic outputs is not yet supported** - #4364.

## Approach 2
Our next instinct was to leverage asset sensors, which should canonically satisfy this use case.
### Implementation
```
from dagster import Array, AssetKey, AssetMaterialization, Field, MetadataValue, Output, RunRequest, asset_sensor, job, op

from mld1_dagster_workflows.python_modules.utils.io import frozendict_to_dict

CONFIG2_SCHEMA = {"run_name": Field(str), "param1": Field(int), "param2": Field(int)}
CONFIG1_SCHEMA = {"run_name": Field(str), "param1": Field(int), "param2": Field(int), "downstream": Field(Array(CONFIG2_SCHEMA))}
CONFIG0_SCHEMA = {"downstream": Field(Array(CONFIG1_SCHEMA))}

@op(config_schema=CONFIG0_SCHEMA)
def run_complete_workflow(context):
    context.log.info("In mock_op1.")
    context.log.info(context.op_config)
    
    yield AssetMaterialization(
        asset_key=AssetKey("1"),
        metadata={
            "native_run_id": context.run_id,
            "upstream_run_id": context.run_id,
            "runs": MetadataValue.json({"runs": frozendict_to_dict(context.op_config["downstream"])})
        },
    )
    yield Output(None)


@op(config_schema=CONFIG1_SCHEMA)
def mock_op1(context):
    context.log.info("In mock_op1.")
    context.log.info(context.op_config)

    yield AssetMaterialization(
        asset_key=AssetKey("1-2"),
        metadata={
            "native_run_id": context.get_tag("native_run_id"),
            "upstream_run_id": context.run_id,
            "runs": MetadataValue.json({"runs": frozendict_to_dict(context.op_config["downstream"])})
        },
    )
    yield Output(None)

@op(config_schema=CONFIG2_SCHEMA)
def mock_op2(context):
    context.log.info("In mock_op2.")
    context.log.info(context.op_config)

    # yield another AssetMaterialization in a more realistic job

@job
def complete_job():
    run_complete_workflow()

@job
def mock_job1():
    mock_op1()

@job
def mock_job2():
    mock_op2()

@asset_sensor(asset_key=AssetKey("1-2"), job=mock_job2, minimum_interval_seconds=15)
def op_1_2_asset_sensor(context, asset_event):
    asset_materialization = asset_event.dagster_event.event_specific_data.materialization
    # this is a pretty awkward way to retrieve asset metadata
    metadata = {entry.label: entry.entry_data for entry in asset_materialization.metadata_entries}
    assert all(k in metadata for k in ["native_run_id", "upstream_run_id", "runs"])
    for i, run in enumerate(metadata["runs"].data["runs"]):
        print(f'\nAttempting to submit run_key {metadata["upstream_run_id"].text}-{i}\n')
        yield RunRequest(
            run_key=f'{metadata["upstream_run_id"].text}-{i}',
            run_config={
                "ops": {
                    "mock_op2": {
                        "config": run
                    }
                }
            },
            tags={
                "run_name": run["run_name"],
                "native_run_id": metadata["native_run_id"].text,
                "upstream_run_id": metadata["upstream_run_id"].text,
            }
        )

@asset_sensor(asset_key=AssetKey("1"), job=mock_job1, minimum_interval_seconds=15)
def op_1_asset_sensor(context, asset_event):
    asset_materialization = asset_event.dagster_event.event_specific_data.materialization
    metadata = {entry.label: entry.entry_data for entry in asset_materialization.metadata_entries}
    assert all(k in metadata for k in ["native_run_id", "upstream_run_id", "runs"])
    for i, run in enumerate(metadata["runs"].data["runs"]):
        print(f'\nAttempting to submit run_key {metadata["upstream_run_id"].text}-{i}\n')
        yield RunRequest(
            run_key=f'{metadata["upstream_run_id"].text}-{i}',
            run_config={
                "ops": {
                    "mock_op1": {
                        "config": run
                    }
                }
            },
            tags={
                "run_name": run["run_name"],
                "native_run_id": metadata["native_run_id"].text,
                "upstream_run_id": metadata["upstream_run_id"].text
            }
        )
```
However, we run into the following where **asset sensors only yield RunRequests for the single last materialized asset (even if several have been materialized between ticks** - #5699. This prevents downstream execution from being fully reliable.
### Example Run
![image](https://user-images.githubusercontent.com/17210429/163223540-981676fa-034f-4b3f-8708-6e092a847e62.png)
(We see here that only 4 of the expected 6 mock_job2 sub-jobs were scheduled.)

## Summary
Both approaches are missing a key feature to fully support our use case.

# Reporting and visualization

The second piece of this use case is to report on the entire job status. Following on from approach 2, a limitation of using chains of sensored-triggered jobs is that one cannot easily visualize the complete DAG in a single place (as if it were an unbroken dagster job). Using the native_run_id and upstream_run_ids defined above however, we could reconstruct the complete DAG using info from a GraphQL query, inject status information, and send this information as part of the eventual slack message body (as well as add links to specific sub-job executions in dagit for reference).
## GraphQL result
```
# query filter: filter: {tags: [{key: "native_run_id", value: "9a041f6d-d58e-4c91-8758-79fefe023aa1"}]}
{
  "data": {
    "runsOrError": {
      "__typename": "Runs",
      "results": [
        {
          "runId": "b64a418b-6822-4d8d-9bbe-7e95ae0d25ba",
          "jobName": "mock_job2",
          "status": "SUCCESS",
          "assets": [],
          "tags": [
            {
              "key": "dagster/run_key",
              "value": "1cb77bca-8da2-4177-95b1-d4e9915cc1e7-1"
            },
            {
              "key": "dagster/sensor_name",
              "value": "op_1_2_asset_sensor"
            },
            {
              "key": "native_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            },
            {
              "key": "run_name",
              "value": "nav6"
            },
            {
              "key": "upstream_run_id",
              "value": "1cb77bca-8da2-4177-95b1-d4e9915cc1e7"
            }
          ]
        },
        {
          "runId": "1fc582da-c7b1-4328-9042-9918e2373629",
          "jobName": "mock_job2",
          "status": "SUCCESS",
          "assets": [],
          "tags": [
            {
              "key": "dagster/run_key",
              "value": "1cb77bca-8da2-4177-95b1-d4e9915cc1e7-0"
            },
            {
              "key": "dagster/sensor_name",
              "value": "op_1_2_asset_sensor"
            },
            {
              "key": "native_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            },
            {
              "key": "run_name",
              "value": "nav5"
            },
            {
              "key": "upstream_run_id",
              "value": "1cb77bca-8da2-4177-95b1-d4e9915cc1e7"
            }
          ]
        },
        {
          "runId": "11c1fc4a-1536-4bd2-8172-3e95b3ef2a77",
          "jobName": "mock_job2",
          "status": "SUCCESS",
          "assets": [],
          "tags": [
            {
              "key": "dagster/run_key",
              "value": "a6e70f72-0a83-47a1-81ee-c2f7bb07a7b2-1"
            },
            {
              "key": "dagster/sensor_name",
              "value": "op_1_2_asset_sensor"
            },
            {
              "key": "native_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            },
            {
              "key": "run_name",
              "value": "nav4"
            },
            {
              "key": "upstream_run_id",
              "value": "a6e70f72-0a83-47a1-81ee-c2f7bb07a7b2"
            }
          ]
        },
        {
          "runId": "80b31b84-a34a-476a-a211-17a150a75ed7",
          "jobName": "mock_job2",
          "status": "SUCCESS",
          "assets": [],
          "tags": [
            {
              "key": "dagster/run_key",
              "value": "a6e70f72-0a83-47a1-81ee-c2f7bb07a7b2-0"
            },
            {
              "key": "dagster/sensor_name",
              "value": "op_1_2_asset_sensor"
            },
            {
              "key": "native_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            },
            {
              "key": "run_name",
              "value": "nav3"
            },
            {
              "key": "upstream_run_id",
              "value": "a6e70f72-0a83-47a1-81ee-c2f7bb07a7b2"
            }
          ]
        },
        {
          "runId": "1cb77bca-8da2-4177-95b1-d4e9915cc1e7",
          "jobName": "mock_job1",
          "status": "SUCCESS",
          "assets": [
            {
              "id": "AssetKey(['1-2'])",
              "key": {
                "path": [
                  "1-2"
                ]
              }
            }
          ],
          "tags": [
            {
              "key": "dagster/run_key",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1-2"
            },
            {
              "key": "dagster/sensor_name",
              "value": "op_1_asset_sensor"
            },
            {
              "key": "native_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            },
            {
              "key": "run_name",
              "value": "mfd3"
            },
            {
              "key": "upstream_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            }
          ]
        },
        {
          "runId": "a6e70f72-0a83-47a1-81ee-c2f7bb07a7b2",
          "jobName": "mock_job1",
          "status": "SUCCESS",
          "assets": [
            {
              "id": "AssetKey(['1-2'])",
              "key": {
                "path": [
                  "1-2"
                ]
              }
            }
          ],
          "tags": [
            {
              "key": "dagster/run_key",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1-1"
            },
            {
              "key": "dagster/sensor_name",
              "value": "op_1_asset_sensor"
            },
            {
              "key": "native_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            },
            {
              "key": "run_name",
              "value": "mfd2"
            },
            {
              "key": "upstream_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            }
          ]
        },
        {
          "runId": "616bccdf-91ad-4609-a3d2-90b39f6f1a9f",
          "jobName": "mock_job1",
          "status": "FAILURE",   # set this as FAILURE for example 
          "assets": [
            {
              "id": "AssetKey(['1-2'])",
              "key": {
                "path": [
                  "1-2"
                ]
              }
            }
          ],
          "tags": [
            {
              "key": "dagster/run_key",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1-0"
            },
            {
              "key": "dagster/sensor_name",
              "value": "op_1_asset_sensor"
            },
            {
              "key": "native_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            },
            {
              "key": "run_name",
              "value": "mfd1"
            },
            {
              "key": "upstream_run_id",
              "value": "9a041f6d-d58e-4c91-8758-79fefe023aa1"
            }
          ]
        }
      ]
    }
  }
}
```
Resulting DAG
![report_dag](https://user-images.githubusercontent.com/17210429/163217929-b5002a5e-1f66-4a7a-849e-7554e0010ec5.png)

Due to the bifurcated nature of this job, another complication here is how we identify when the job is complete in order to send out a single unified slack report. In order to identify when the entire job (including all downstream sub-jobs) is completed, we can use a run status sensor which would be run on all jobs. Given the native_run_id and input config, the run status sensor can run another GraphQL query to identify what the complete DAG should look like and compare it against what jobs/DAG nodes are complete, in-progress, and failed. So, on completion of each sub-job, we would check whether the parent-job the sub-job is part of has hit a terminal state of execution, and send the report if so.

Now assuming #5699 were supported, this method of reporting should probably work, but it feels like a **major over-complication** of what reporting and visualization for a job should look like.

# Summary of feature requests / relevant open issues
- Would there be any way to unify the entire graph structure within the Dagit UI s.t. users could see the flow of success/fails as if it were executed as a single, unsplit job? If so, DAG visualization would be greatly improved and reporting would be as simple as making a single run-status sensor which doesn't involve any additional tagging or graph logic like we'd have above.

Any chance we could revisit the following?
- #5699 - Supporting this would directly unblock the second approach specified above.
- #4364 - Supporting this would unblock the first approach of using nested subgraphs.
- #7384 (lower priority quality of life change)

**Comment by Phazure** [comment]
@yuhan @sryza if you guys had any thoughts. A bit of a long post and apologies in advance if any of the above is confusing (happy to clarify anything as needed), but the main questions are really:
1. What should the best practice be for running very runtime-dynamic jobs? My understanding is the canonical use of asset-triggered sensors doesn't fully work because of #5699, and using many levels of nested subgraphs doesn't fully work because of #4364.
2. How should one set up reporting for a complete dagster job for which ops failed / succeeded, including the downstream jobs that may be triggered separately (if using sensor-triggered child jobs)?

**Comment by sryza** [comment]
Hi @Phazure - thanks for this detailed write-up. I've been on vacation, so just getting to this now.

I'm sorry to say that I think the short answer is that the dynamic functionality you're describing is not something we're prioritizing to support in the near future.

We are interested in getting to https://github.com/dagster-io/dagster/issues/4364 soon. However, even when that is done, users will still be expected to specify the number of levels of nesting at the time they define their graphs.  I.e. ops will be dynamic "horizontally", but not "vertically".

If I were in your position, one option I would consider would be to tools like Ray that support dynamic runtime topology as a first-class citizen.

> However, we run into the following where asset sensors only yield RunRequests for the single last materialized asset (even if several have been materialized between ticks - https://github.com/dagster-io/dagster/issues/5699. This prevents downstream execution from being fully reliable.

I don't think it would be difficult to basically copy the asset sensor implementation and modify it to yield a RunRequest for each last materialized event.

**Reply by zigabostner** [reply]
Hi @sryza, this is not directly related to the OPs question, but are you maybe aware of more tools like Ray, that support dynamic runtime topology?

**Reply by sryza** [reply]
@zigabostner - yes, Ray could be a good tool for supporting this kind of dynamic runtime topology.

**Comment by geoHeil** [comment]
https://github.com/asanoop24/dagger/tree/master is perhaps relevant in this discussion
